README.md
# PostCSS Trig [<img src="https://postcss.github.io/postcss/logo.svg" alt="PostCSS" width="90" height="90" align="right">][postcss]

[![NPM Version][npm-img]][npm-url]
[![Build Status][cli-img]][cli-url]
[![Support Chat][git-img]][git-url]

[PostCSS Trig] lets you use trigonometry functions in CSS.

These functions include `acos`, `asin`, `atan`, `atan2`, `cos`, `hypot`, `pow`,
`sin`, `sqrt`, and `tan`.

```pcss
:root {
  --acos: acos(0.8);
  --asin: asin(0.6);
  --atan: atan(0.8);
  --atan2: atan2(5, 5);
  --cos: cos(20);
  --cos-deg: cos(20deg);
  --hypot: hypot(3, 4);
  --pow: pow(4, 4);
  --pow-rem: pow(4rem, 4);
  --sin: sin(20);
  --sin-deg: sin(20deg);
  --sqrt: sqrt(81);
  --sqrt-rem: sqrt(81rem);
  --tan: tan(20);
  --tan-deg: tan(20deg);
}


/* becomes */

:root {
  --acos: 0.6435;
  --asin: 0.6435;
  --atan: 0.9273;
  --atan2: 0.7854;
  --cos: 0.40808;
  --cos-deg: 0.93969;
  --hypot: 5;
  --pow: 256;
  --pow-rem: 256rem;
  --sin: 0.91295;
  --sin-deg: 0.34202;
  --sqrt: 9;
  --sqrt-rem: 9rem;
  --tan: 2.23716;
  --tan-deg: 0.36397;
}
```

When necessary, more complex fallbacks are used to support real-time values.

```pcss
:root {
  --cos-calc: cos(calc(20 * 2));
  --cos-var: cos(var(--angle));
  --pow-calc: pow(calc(4 * 2), 4);
  --pow-var: pow(var(--number), 4);
  --sin-calc: sin(calc(20 * 2));
  --sin-var: sin(var(--angle));
  --sqrt-calc: sqrt(calc(81 * 4));
  --sqrt-var: sqrt(var(--number));
  --tan-calc: tan(calc(20 * 2));
  --tan-var: tan(var(--angle));
}

/* becomes */

:root {
  --cos-calc: calc(1 - ((20 * 2) * (20 * 2) /  2 ) + ((20 * 2) * (20 * 2) * (20 * 2) * (20 * 2) /  24 ) - ((20 * 2) * (20 * 2) * (20 * 2) * (20 * 2) * (20 * 2) * (20 * 2) /  720 ) + ((20 * 2) * (20 * 2) * (20 * 2) * (20 * 2) * (20 * 2) * (20 * 2) * (20 * 2) * (20 * 2) /  40320 ));
  --cos-var: calc(1 - (var(--angle) * var(--angle) /  2 ) + (var(--angle) * var(--angle) * var(--angle) * var(--angle) /  24 ) - (var(--angle) * var(--angle) * var(--angle) * var(--angle) * var(--angle) * var(--angle) /  720 ) + (var(--angle) * var(--angle) * var(--angle) * var(--angle) * var(--angle) * var(--angle) * var(--angle) * var(--angle) /  40320 ));
  --pow-calc: pow(calc(4 * 2), 4);
  --pow-var: pow(var(--number), 4);
  --sin-calc: calc((20 * 2) - ((20 * 2) * (20 * 2) * (20 * 2) /  6 ) + ((20 * 2) * (20 * 2) * (20 * 2) * (20 * 2) * (20 * 2) /  120 ) - ((20 * 2) * (20 * 2) * (20 * 2) * (20 * 2) * (20 * 2) * (20 * 2) * (20 * 2) /  5040 ) + ((20 * 2) * (20 * 2) * (20 * 2) * (20 * 2) * (20 * 2) * (20 * 2) * (20 * 2) * (20 * 2) * (20 * 2) /  362880 ));
  --sin-var: calc(var(--angle) - (var(--angle) * var(--angle) * var(--angle) /  6 ) + (var(--angle) * var(--angle) * var(--angle) * var(--angle) * var(--angle) /  120 ) - (var(--angle) * var(--angle) * var(--angle) * var(--angle) * var(--angle) * var(--angle) * var(--angle) /  5040 ) + (var(--angle) * var(--angle) * var(--angle) * var(--angle) * var(--angle) * var(--angle) * var(--angle) * var(--angle) * var(--angle) /  362880 ));
  --sqrt-calc: calc(((((((((((81 * 4) + (81 * 4) / (81 * 4)) / 2) + (81 * 4) / (((81 * 4) + (81 * 4) / (81 * 4)) / 2)) / 2) + (81 * 4) / (((((81 * 4) + (81 * 4) / (81 * 4)) / 2) + (81 * 4) / (((81 * 4) + (81 * 4) / (81 * 4)) / 2)) / 2)) / 2) + (81 * 4) / (((((((81 * 4) + (81 * 4) / (81 * 4)) / 2) + (81 * 4) / (((81 * 4) + (81 * 4) / (81 * 4)) / 2)) / 2) + (81 * 4) / (((((81 * 4) + (81 * 4) / (81 * 4)) / 2) + (81 * 4) / (((81 * 4) + (81 * 4) / (81 * 4)) / 2)) / 2)) / 2)) / 2) + (81 * 4) / (((((((((81 * 4) + (81 * 4) / (81 * 4)) / 2) + (81 * 4) / (((81 * 4) + (81 * 4) / (81 * 4)) / 2)) / 2) + (81 * 4) / (((((81 * 4) + (81 * 4) / (81 * 4)) / 2) + (81 * 4) / (((81 * 4) + (81 * 4) / (81 * 4)) / 2)) / 2)) / 2) + (81 * 4) / (((((((81 * 4) + (81 * 4) / (81 * 4)) / 2) + (81 * 4) / (((81 * 4) + (81 * 4) / (81 * 4)) / 2)) / 2) + (81 * 4) / (((((81 * 4) + (81 * 4) / (81 * 4)) / 2) + (81 * 4) / (((81 * 4) + (81 * 4) / (81 * 4)) / 2)) / 2)) / 2)) / 2)) / 2);
  --sqrt-var: calc((((((((((var(--number) + var(--number) / var(--number)) / 2) + var(--number) / ((var(--number) + var(--number) / var(--number)) / 2)) / 2) + var(--number) / ((((var(--number) + var(--number) / var(--number)) / 2) + var(--number) / ((var(--number) + var(--number) / var(--number)) / 2)) / 2)) / 2) + var(--number) / ((((((var(--number) + var(--number) / var(--number)) / 2) + var(--number) / ((var(--number) + var(--number) / var(--number)) / 2)) / 2) + var(--number) / ((((var(--number) + var(--number) / var(--number)) / 2) + var(--number) / ((var(--number) + var(--number) / var(--number)) / 2)) / 2)) / 2)) / 2) + var(--number) / ((((((((var(--number) + var(--number) / var(--number)) / 2) + var(--number) / ((var(--number) + var(--number) / var(--number)) / 2)) / 2) + var(--number) / ((((var(--number) + var(--number) / var(--number)) / 2) + var(--number) / ((var(--number) + var(--number) / var(--number)) / 2)) / 2)) / 2) + var(--number) / ((((((var(--number) + var(--number) / var(--number)) / 2) + var(--number) / ((var(--number) + var(--number) / var(--number)) / 2)) / 2) + var(--number) / ((((var(--number) + var(--number) / var(--number)) / 2) + var(--number) / ((var(--number) + var(--number) / var(--number)) / 2)) / 2)) / 2)) / 2)) / 2);
  --tan-calc: calc((20 * 2) + ((1 / 3) * ((20 * 2) * (20 * 2) * (20 * 2))) + ((2 / 15) * ((20 * 2) * (20 * 2) * (20 * 2) * (20 * 2) * (20 * 2))) + ((17 / 315) * ((20 * 2) * (20 * 2) * (20 * 2) * (20 * 2) * (20 * 2) * (20 * 2) * (20 * 2))));
  --tan-var: calc(var(--angle) + ((1 / 3) * (var(--angle) * var(--angle) * var(--angle))) + ((2 / 15) * (var(--angle) * var(--angle) * var(--angle) * var(--angle) * var(--angle))) + ((17 / 315) * (var(--angle) * var(--angle) * var(--angle) * var(--angle) * var(--angle) * var(--angle) * var(--angle))));
}
```

## Usage

Add [PostCSS Trig] to your project:

```bash
npm install postcss-trig --save-dev
```

Use [PostCSS Trig] to process your CSS:

```js
const postcssTrig = require('postcss-trig');

postcssTrig.process(YOUR_CSS /*, processOptions, pluginOptions */);
```

Or use it as a [PostCSS] plugin:

```js
const postcss = require('postcss');
const postcssTrig = require('postcss-trig');

postcss([
  postcssTrig(/* pluginOptions */)
]).process(YOUR_CSS /*, processOptions */);
```

[PostCSS Trig] runs in all Node environments, with special instructions for:

| [Node](INSTALL.md#node) | [PostCSS CLI](INSTALL.md#postcss-cli) | [Webpack](INSTALL.md#webpack) | [Create React App](INSTALL.md#create-react-app) | [Gulp](INSTALL.md#gulp) | [Grunt](INSTALL.md#grunt) |
| --- | --- | --- | --- | --- | --- |

## Options

### precision

The `precision` option determines how many decimal places of precision will be
preserved after transformations. By default, this precision is up to `5` places.

[cli-img]: https://img.shields.io/travis/csstools/postcss-trig.svg
[cli-url]: https://travis-ci.org/csstools/postcss-trig
[git-img]: https://img.shields.io/badge/support-chat-blue.svg
[git-url]: https://gitter.im/postcss/postcss
[npm-img]: https://img.shields.io/npm/v/postcss-trig.svg
[npm-url]: https://www.npmjs.com/package/postcss-trig

[PostCSS]: https://github.com/postcss/postcss
[PostCSS Trig]: https://github.com/csstools/postcss-trig


CHANGELOG.md
# Changes to PostCSS Trig

### 2.0.0 (May 12, 2019)

- Updated: `postcss` to 7.0.16 (patch)
- Updated: `postcss-values-parser` to 3.0.4 (major)
- Updated: Node 8+ compatibility (major)

### 1.0.0 (February 28, 2019)

- Initial version


src/index.js
import postcss from 'postcss';
import valuesParser from 'postcss-values-parser';
import {
	transformAcosFunction,
	transformAsinFunction,
	transformAtanFunction,
	transformAtan2Function,
	transformCosFunction,
	transformHypotFunction,
	transformPowFunction,
	transformSinFunction,
	transformSqrtFunction,
	transformTanFunction
} from './lib/transforms';

export default postcss.plugin('postcss-trig', opts => {
	const precision = 'precision' in Object(opts) ? Number(opts.precision) || 0 : 5;

	return root => {
		root.walkDecls(decl => {
			if (anyTrigFunctionRegExp.test(decl.value)) {
				const originalValue = decl.value;

				const valueAST = valuesParser(originalValue).parse();

				valueAST.walkFunctionNodes(functionNode => {
					if (trigFunctionNameRegExp.test(functionNode.value)) {
						const name = functionNode.value.toLowerCase();
						const transform = transformsByName[name];

						transform(functionNode, precision);
					}
				});

				const modifiedValue = String(valueAST);

				if (originalValue !== modifiedValue) {
					decl.value = modifiedValue;
				}
			}
		});
	};
});

const transformsByName = {
	acos: transformAcosFunction,
	asin: transformAsinFunction,
	atan: transformAtanFunction,
	atan2: transformAtan2Function,
	cos: transformCosFunction,
	hypot: transformHypotFunction,
	pow: transformPowFunction,
	sin: transformSinFunction,
	sqrt: transformSqrtFunction,
	tan: transformTanFunction
};

const anyTrigFunctionRegExp = /\b(acos|asin|atan|atan2|cos|hypot|pow|sin|sqrt|tan)\(.*\)/i;
const trigFunctionNameRegExp = /^(acos|asin|atan|atan2|cos|hypot|pow|sin|sqrt|tan)$/i


src/lib/is.js
/* Numbers
/* ========================================================================== */

// return whether the length is valid
export function isLength(node) {
	return Object(node).type === 'number';
}

// return whether the length is valid
export function isAngle(node) {
	return isLength(node) && matchAngleUnit.test(node.unit);
}

// return whether the number is valid
export function isNumber(node) {
	return isLength(node) && node.unit === '';
}

/* Delimiters
/* ========================================================================== */

// return whether the comma is valid
export function isComma(node) {
	return Object(node).type === 'comma';
}

/* Functions
/* ========================================================================== */

export function isFunction(node) {
	return Object(node).type === 'func';
}

export function isCalc(node) {
	return isFunction(node) && matchCalcName.test(node.value);
}

/* Matchers
/* ========================================================================== */

export const matchAngleUnit = /^(deg|grad|rad|turn)$/i;
export const matchCalcName = /^calc$/i;


src/lib/transforms.js
import { isAngle, isCalc, isComma, isFunction, isLength, isNumber, matchCalcName } from './is';
import { newNumber, newOperator, newFunction } from './new';

/* Transforms
/* ========================================================================== */

// return the first set of transformed arguments allowable by the parameters
export function transformArgsByParams(node, params) {
	const nodes = (Object(node).nodes || []).slice(1, -1);

	return params.map(param => nodes.map(
		(childNode, index) => typeof param[index] === 'function' ? param[index](childNode) : null
	).filter(child => typeof child !== 'boolean')).filter(param => param.every(
		result => result !== null
	))[0] || [];
}

// return a transformed angle
export function transformAngleOrNumberNodeIntoNumberNode(node) {
	if (isAngle(node) || isNumber(node)) {
		const unit = node.unit.toLowerCase();

		if (unit in angleMultipliersByUnit) {
			node.value = String(Number(node.value) * angleMultipliersByUnit[unit]);

			node.unit = '';
		}

		return node;
	}

	return null;
}

function transformCalcAddDivideBy(node, number) {
	if (isCalc(node)) {
		node.nodes.splice(-1, 0, newOperator('/'), newOperator(number));
	}

	return node;
}

export function getFunctionNode(node) {
	return isFunction(node) ? node : null;
}

export function getLengthNode(node) {
	return isLength(node) ? node : null;
}

export function getNumber(node) {
	return isNumber(node) ? Number(node.value) : null;
}

export function getNumberNode(node) {
	return isNumber(node) ? node : null;
}

export function uncalcNode(node) {
	if (isCalc(node)) {
		node.value = node.value.replace(matchCalcName, '');
	}

	return node;
}

/* Transform acos()
/* ========================================================================== */

export function transformAcosFunction(functionNode, precision) {
	const [radianNode] = transformArgsByParams(functionNode, [
		// <angle> | <number>
		[transformAngleOrNumberNodeIntoNumberNode]
	]);

	if (isNumber(radianNode)) {
		const numberNode = getAcosNumberNodeFromNumberNode(radianNode, precision);

		if (numberNode.value !== 'NaN') {
			functionNode.replaceWith(numberNode);
		}
	}
}

export function getAcosNumberNodeFromNumberNode(numberNode, precision) {
	const acos = usePrecision(Math.acos(numberNode.value), precision);

	return clone(numberNode, { value: String(acos) });
}

/* Transform asin()
/* ========================================================================== */

export function transformAsinFunction(functionNode, precision) {
	const [radianNode] = transformArgsByParams(functionNode, [
		// <angle> | <number>
		[transformAngleOrNumberNodeIntoNumberNode]
	]);

	if (isNumber(radianNode)) {
		const numberNode = getAsinNumberNodeFromNumberNode(radianNode, precision);

		if (numberNode.value !== 'NaN') {
			functionNode.replaceWith(numberNode);
		}
	}
}

export function getAsinNumberNodeFromNumberNode(numberNode, precision) {
	const asin = usePrecision(Math.asin(numberNode.value), precision);

	return clone(numberNode, { value: String(asin) });
}

/* Transform asin()
/* ========================================================================== */

export function transformAtanFunction(functionNode, precision) {
	const [radianNode] = transformArgsByParams(functionNode, [
		// <angle> | <number>
		[transformAngleOrNumberNodeIntoNumberNode]
	]);

	if (isNumber(radianNode)) {
		const numberNode = getAtanNumberNodeFromNumberNode(radianNode, precision);

		if (numberNode.value !== 'NaN') {
			functionNode.replaceWith(numberNode);
		}
	}
}

export function getAtanNumberNodeFromNumberNode(numberNode, precision) {
	const atan = usePrecision(Math.asin(numberNode.value), precision);

	return clone(numberNode, { value: String(atan) });
}

/* Transform atan2()
/* ========================================================================== */

export function transformAtan2Function(functionNode, precision) {
	const [coordinateY, coordinateX] = transformArgsByParams(functionNode, [
		// <number>, <number>
		[getNumber, isComma, getNumber]
	]);

	if (typeof coordinateY === 'number' && typeof coordinateX === 'number') {
		const numberNode = getAtan2NumberNodeFromCoordinates(coordinateY, coordinateX, precision);

		if (numberNode.value !== 'NaN') {
			functionNode.replaceWith(numberNode);
		}
	}
}

export function getAtan2NumberNodeFromCoordinates(coordinateY, coordinateX, precision) {
	const atan2 = usePrecision(Math.atan2(coordinateY, coordinateX), precision);

	return newNumber(atan2);
}

/* Transform cos()
/* ========================================================================== */

export function transformCosFunction(cosFunctionNode, precision) {
	const [radianNode] = transformArgsByParams(cosFunctionNode, [
		// <angle> | <number>
		[transformAngleOrNumberNodeIntoNumberNode],
		// <function>
		[getFunctionNode]
	]);

	if (isNumber(radianNode)) {
		const numberNode = getCosNumberNodeFromNumberNode(radianNode, precision);

		if (numberNode.value !== 'NaN') {
			cosFunctionNode.replaceWith(numberNode);
		}
	} else if (isFunction(radianNode)) {
		cosFunctionNode.replaceWith(getCosFunctionNodeFromFunctionNode(radianNode));
	}
}

export function getCosNumberNodeFromNumberNode(numberNode, precision) {
	const cos = usePrecision(Math.cos(numberNode.value), precision);

	return clone(numberNode, { value: String(cos) });
}

export function getCosFunctionNodeFromFunctionNode(functionNode) {
	const cosTermNode1 = newNumber(1);
	const cosTermNode2 = transformCalcAddDivideBy(getPowCalcFunctionNodeFromFunctionNode(functionNode, 2), 2);
	const cosTermNode3 = transformCalcAddDivideBy(getPowCalcFunctionNodeFromFunctionNode(functionNode, 4), 24);
	const cosTermNode4 = transformCalcAddDivideBy(getPowCalcFunctionNodeFromFunctionNode(functionNode, 6), 720);
	const cosTermNode5 = transformCalcAddDivideBy(getPowCalcFunctionNodeFromFunctionNode(functionNode, 8), 40320);
	const calcCosNode = newFunction('calc', ...[
		uncalcNode(cosTermNode1),
		newOperator('-'),
		uncalcNode(cosTermNode2),
		newOperator('+'),
		uncalcNode(cosTermNode3),
		newOperator('-'),
		uncalcNode(cosTermNode4),
		newOperator('+'),
		uncalcNode(cosTermNode5)
	]);

	return calcCosNode;
}

/* Transform hypot()
/* ========================================================================== */

export function transformHypotFunction(hypotFunctionNode, precision) {
	const hypotNodes = hypotFunctionNode.nodes.slice(1, -1);

	const isValidHypotArgs = hypotNodes.every((hypotNode, index) => index % 2 ? isComma(hypotNode) : isNumber(hypotNode));

	if (isValidHypotArgs) {
		const numberNode = getHypotNumberNodeFromHypotNodes(hypotNodes, precision);

		if (numberNode.value !== 'NaN') {
			hypotFunctionNode.replaceWith(numberNode);
		}
	}
}

export function getHypotNumberNodeFromHypotNodes(hypotNodes, precision) {
	const hypotNodesValues = hypotNodes.filter((hypotNode, index) => !(index % 2)).map(hypotNode => hypotNode.value);
	const hypot = usePrecision(Math.hypot.apply(Math, hypotNodesValues), precision);
	const hypotNode = newNumber(hypot);

	return hypotNode;
}

/* Transform pow()
/* ========================================================================== */

export function transformPowFunction(powFunctionNode, precision) {
	const [baseNode, exponent] = transformArgsByParams(powFunctionNode, [
		// <length>, <number>
		[getLengthNode, isComma, getNumber],
		// <function>, <number>
		[getFunctionNode, isComma, getNumber]
	]);

	if (isLength(baseNode) && typeof exponent === 'number') {
		const lengthNode = getPowLengthNodeFromLengthNode(baseNode, exponent, precision);

		if (lengthNode.value !== 'NaN') {
			powFunctionNode.replaceWith(lengthNode);
		}
	} else if (isFunction(baseNode) && typeof exponentNumber === 'number') {
		powFunctionNode.replaceWith(getPowCalcFunctionNodeFromFunctionNode(baseNode, exponent));
	}
}

export function getPowLengthNodeFromLengthNode(lengthNode, exponent, precision) {
	const base = Number(lengthNode.value);
	const pow = usePrecision(Math.pow(base, exponent), precision);

	return clone(lengthNode, { value: String(pow) });
}

export function getPowCalcFunctionNodeFromFunctionNode(functionNode, exponent) {
	const functionCloneNode = uncalcNode(clone(functionNode));
	const calcSumNodes = [];
	let i = -1;

	while (++i < exponent) {
		calcSumNodes.push(clone(functionCloneNode));

		if (i + 1 < exponent) {
			calcSumNodes.push(newOperator('*'));
		}
	}

	return newFunction('calc', ...calcSumNodes);
}

/* Transform sin()
/* ========================================================================== */

export function transformSinFunction(sinFunctionNode, precision) {
	const [radianNode] = transformArgsByParams(sinFunctionNode, [
		// <angle> | <number>
		[transformAngleOrNumberNodeIntoNumberNode],
		// <function>
		[getFunctionNode]
	]);

	if (isNumber(radianNode)) {
		const numberNode = getSinNumberNodeFromNumberNode(radianNode, precision);

		if (numberNode.value !== 'NaN') {
			sinFunctionNode.replaceWith(numberNode);
		}
	} else if (isFunction(radianNode)) {
		sinFunctionNode.replaceWith(getSinCalcFunctionNodeFromFunctionNode(radianNode));
	}
}

export function getSinNumberNodeFromNumberNode(numberNode, precision) {
	const sin = usePrecision(Math.sin(numberNode.value), precision);

	return clone(numberNode, { value: String(sin) });
}

export function getSinCalcFunctionNodeFromFunctionNode(functionNode) {
	const sinTermNode1 = clone(functionNode);
	const sinTermNode2 = transformCalcAddDivideBy(getPowCalcFunctionNodeFromFunctionNode(functionNode, 3), 6);
	const sinTermNode3 = transformCalcAddDivideBy(getPowCalcFunctionNodeFromFunctionNode(functionNode, 5), 120);
	const sinTermNode4 = transformCalcAddDivideBy(getPowCalcFunctionNodeFromFunctionNode(functionNode, 7), 5040);
	const sinTermNode5 = transformCalcAddDivideBy(getPowCalcFunctionNodeFromFunctionNode(functionNode, 9), 362880);
	const calcSinNode = newFunction('calc', ...[
		uncalcNode(sinTermNode1),
		newOperator('-'),
		uncalcNode(sinTermNode2),
		newOperator('+'),
		uncalcNode(sinTermNode3),
		newOperator('-'),
		uncalcNode(sinTermNode4),
		newOperator('+'),
		uncalcNode(sinTermNode5)
	]);

	return calcSinNode;
}

/* Transform sqrt()
/* ========================================================================== */

export function transformSqrtFunction(sqrtFunctionNode, precision) {
	const [baseNode] = transformArgsByParams(sqrtFunctionNode, [
		// <length>, <number>
		[getLengthNode],
		// <function>, <number>
		[getFunctionNode]
	]);

	if (isLength(baseNode)) {
		const sqrtNumberNode = getSqrtLengthNodeFromLengthNode(baseNode, precision);

		if (sqrtNumberNode.value !== 'NaN') {
			sqrtFunctionNode.replaceWith(sqrtNumberNode);
		}
	} else if (isFunction(baseNode)) {
		sqrtFunctionNode.replaceWith(getSqrtCalcFunctionNodeFromFunctionNode(baseNode));
	}
}

export function getSqrtLengthNodeFromLengthNode(lengthNode, precision) {
	const sqrt = usePrecision(Math.sqrt(lengthNode.value), precision);

	return clone(lengthNode, { value: String(sqrt) });
}

export function getSqrtCalcFunctionNodeFromFunctionNode(functionNode) {
	const sqrtGuess1 = getSqrtGuessCalcFunctionNodeFromFunctionNode(functionNode, functionNode);
	const sqrtGuess2 = getSqrtGuessCalcFunctionNodeFromFunctionNode(sqrtGuess1, functionNode);
	const sqrtGuess3 = getSqrtGuessCalcFunctionNodeFromFunctionNode(sqrtGuess2, functionNode);
	const sqrtGuess4 = getSqrtGuessCalcFunctionNodeFromFunctionNode(sqrtGuess3, functionNode);
	const sqrtGuess5 = getSqrtGuessCalcFunctionNodeFromFunctionNode(sqrtGuess4, functionNode);

	return sqrtGuess5;
}

export function getSqrtGuessCalcFunctionNodeFromFunctionNode(guessFunctionNode, originalFunctionNode) {
	return newFunction('calc', ...[
		newFunction('', ...[
			...[
				uncalcNode(clone(guessFunctionNode)),
				newOperator('+'),
				uncalcNode(clone(originalFunctionNode)),
				newOperator('/'),
				uncalcNode(clone(guessFunctionNode))
			]
		]),
		newOperator('/'),
		newNumber(2)
	]);
}

/* Transform tan()
/* ========================================================================== */

export function transformTanFunction(tanFunctionNode, precision) {
	const [radianNode] = transformArgsByParams(tanFunctionNode, [
		// <angle> | <number>
		[transformAngleOrNumberNodeIntoNumberNode],
		// <function>
		[getFunctionNode]
	]);

	if (isNumber(radianNode)) {
		const numberNode = getTanNumberNodeFromNumberNode(radianNode, precision);

		if (numberNode.value !== 'NaN') {
			tanFunctionNode.replaceWith(numberNode);
		}
	} else if (isFunction(radianNode)) {
		tanFunctionNode.replaceWith(getTanCalcFunctionNodeFromFunctionNode(radianNode));
	}
}

export function getTanNumberNodeFromNumberNode(numberNode, precision) {
	const tan = usePrecision(Math.tan(numberNode.value), precision);

	return clone(numberNode, { value: String(tan) });
}

export function getTanCalcFunctionNodeFromFunctionNode(functionNode) {
	const tanTermNode1 = clone(functionNode);
	const tanTermNode2 = newFunction('calc', ...[
		newFunction('', ...[
			newNumber(1), newOperator('/'), newNumber(3)
		]),
		newOperator('*'),
		...getPowCalcFunctionNodeFromFunctionNode(functionNode, 3).nodes
	]);
	const tanTermNode3 = newFunction('calc', ...[
		newFunction('', ...[
			newNumber(2), newOperator('/'), newNumber(15)
		]),
		newOperator('*'),
		...getPowCalcFunctionNodeFromFunctionNode(functionNode, 5).nodes
	]);
	const tanTermNode4 = newFunction('calc', ...[
		newFunction('', ...[
			newNumber(17), newOperator('/'), newNumber(315)
		]),
		newOperator('*'),
		...getPowCalcFunctionNodeFromFunctionNode(functionNode, 7).nodes
	]);
	const calcTanNode = newFunction('calc', ...[
		uncalcNode(tanTermNode1),
		newOperator('+'),
		uncalcNode(tanTermNode2),
		newOperator('+'),
		uncalcNode(tanTermNode3),
		newOperator('+'),
		uncalcNode(tanTermNode4)
	]);

	return calcTanNode;
}

/* Conversions
/* ========================================================================== */

const angleMultipliersByUnit = {
	'': 1,
	deg: Math.PI / 180,
	grad: Math.PI / 200,
	rad: 1,
	turn: 2 * Math.PI
};

/* Clone
/* ========================================================================== */

function clone(node, overrides) {
	return normalize(node, node.clone(overrides));
}

function normalize(node, cloneNode) {
	if (node.raws) {
		for (const raw in node.raws) {
			cloneNode.raws[raw] = node.raws[raw];
		}
	}

	if (node.nodes) {
		for (const index in node.nodes) {
			normalize(node.nodes[index], cloneNode.nodes[index]);
		}
	}

	return cloneNode;
}

function usePrecision(number, precision) {
	return Number(`${Math.round(`${number}e${precision}`)}e-${precision}`);
}
